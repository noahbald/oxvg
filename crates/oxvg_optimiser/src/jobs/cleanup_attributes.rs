use serde::Deserialize;

use crate::Job;

#[derive(Deserialize, Default)]
pub struct CleanupAttributes {
    newlines: Option<bool>,
    trim: Option<bool>,
    spaces: Option<bool>,
}

impl Job for CleanupAttributes {
    fn from_configuration(value: serde_json::Value) -> Self {
        serde_json::from_value(value).unwrap_or_default()
    }

    fn run(&self, node: &rcdom::Node) {
        use rcdom::NodeData::Element;

        let Element { attrs, .. } = &node.data else {
            return;
        };
        let attrs = &mut *attrs.borrow_mut();

        for attr in attrs.iter_mut() {
            if matches!(self.newlines, Some(true)) {
                attr.value = attr.value.replace('\n', " ").into();
            }
            if matches!(self.trim, Some(true)) {
                attr.value = attr.value.trim().into();
            }
            if matches!(self.spaces, Some(true)) {
                attr.value = attr
                    .value
                    .split_whitespace()
                    .collect::<Vec<_>>()
                    .join(" ")
                    .into();
            }
        }
    }
}

#[test]
fn cleanup_attributes() -> Result<(), &'static str> {
    use rcdom::NodeData::Element;
    use xml5ever::{
        driver::{parse_document, XmlParseOpts},
        tendril::TendrilSink,
    };

    let dom: rcdom::RcDom = parse_document(rcdom::RcDom::default(), XmlParseOpts::default()).one(
        r#"<svg class="  foo  bar 
        baz"></svg>"#,
    );
    let root = &*dom.document.children.borrow()[0];
    let job = CleanupAttributes {
        newlines: Some(true),
        trim: Some(true),
        spaces: Some(true),
    };

    job.run(root);
    let attrs = match &root.data {
        Element { attrs, .. } => attrs,
        _ => Err("Unexpected document structure")?,
    };
    let attrs = &*attrs.borrow();
    assert_eq!(
        attrs.first().map(|attr| &attr.value),
        Some(&"foo bar baz".into())
    );

    Ok(())
}
