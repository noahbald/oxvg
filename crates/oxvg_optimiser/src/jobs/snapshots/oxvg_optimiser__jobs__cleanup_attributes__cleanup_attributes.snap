---
source: crates/oxvg_optimiser/src/jobs/cleanup_attributes.rs
expression: "test_config(r#\"{ \"cleanupAttributes\": {\n            \"newlines\": true,\n            \"trim\": true,\n            \"spaces\": true\n        } }\"#,\n        Some(r#\"<svg xmlns=\"  http://www.w3.org/2000/svg\n  \" attr=\"a      b\" attr2=\"a\nb\">\n    test\n</svg>\"#))?"
---
<svg xmlns="  http://www.w3.org/2000/svg
  " attr="a b" attr2="a b">
    test
</svg>
